pgjdbc/src/main/java/org/postgresql/jdbc/TimestampUtils.java

```diff
@@ -454,6 +456,44 @@ public LocalDateTime toLocalDateTime(String s) throws SQLException {
       return result;
     }
   }
+
+
+  /**
+   * Parse a string and return a ZonedDateTime representing its value.
+   *
+   * @param s The ISO formated date string to parse.
+   * @return null if s is null or a ZonedDateTime of the parsed string s.
+   * @throws SQLException if there is a problem parsing s.
+   */
+  public ZonedDateTime toZonedDateTime(String s) throws SQLException {
+    if (s == null) {
+      return null;
+    }
+
+    int slen = s.length();
+
+    // convert postgres's infinity values to internal infinity magic value
+    if (slen == 8 && s.equals("infinity")) {
+      return ZonedDateTime.of(LocalDateTime.MAX, ZoneId.of("UTC"));
+    }
+
+    if (slen == 9 && s.equals("-infinity")) {
+      return ZonedDateTime.of(LocalDateTime.MIN, ZoneId.of("UTC"));
+    }
+
+    ParsedTimestamp ts = parseBackendTimestamp(s);
+
+    LocalDateTime localDateTime = LocalDateTime.of(ts.year, ts.month, ts.day, ts.hour, ts.minute, ts.second, ts.nanos);
+    ZonedDateTime result = ZonedDateTime.of(localDateTime, ts.tz.getTimeZone().toZoneId());
```

Comment:
This looks odd.
I wonder why ZonedDateTime requires its own implementation while OffsetDateTime does not.
Could ZonedDateTime just wrap OffsetDateTime?

Reply:
Very good question; from the javadocs:
"OffsetDateTime, ZonedDateTime and Instant all store an instant on the time-line to nanosecond precision. Instant is the simplest, simply representing the instant. OffsetDateTime adds to the instant the offset from UTC/Greenwich, which allows the local date-time to be obtained. ZonedDateTime adds full time-zone rules."

So do we need all the timezone rules ?

Reply:
Just refactored it and you are correct it does not require it.

Reply:
>So do we need all the timezone rules ?

That's a good question. I might schedule a meeting on `15 April 2018 Europe/Berlin`.

It might be even better to fail when user tries to insert `Europe/Berlin` instead of silently converting it to `+01:00`.